src/peft/tuners/delora/layer.py

```diff
@@ -0,0 +1,630 @@
+# Copyright 2025-present the HuggingFace Inc. team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+
+import math
+import warnings
+from typing import Any, Literal, Optional
+
+import torch
+import torch.nn as nn
+
+from peft.tuners._buffer_dict import BufferDict
+from peft.tuners.tuners_utils import BaseTunerLayer, check_adapters_to_merge
+
+
+class DeLoRALayer(BaseTunerLayer):
+    # All names of layers that may contain (trainable) adapter weights
+    adapter_layer_names: tuple[str, ...] = ("delora_A", "delora_B", "delora_alpha")
+    # All names of other parameters that may contain adapter-related parameters
+    other_param_names: tuple[str, ...] = (
+        "r",
+        "delora_dropout",
+        # Persisted initial copies and norms (use delora_ prefix so they are saved)
+        "delora_initial_A",
+        "delora_initial_B",
+        "delora_initial_alpha",
+        "delora_w_norm",
+        "delora_initial_w_norm",
+        # Misc flag
+        "use_residual_init",
+    )
+
+    def __init__(self, base_layer: nn.Module, **kwargs) -> None:
+        self.base_layer = base_layer
+        self.r = {}
+        self.delora_dropout = nn.ModuleDict({})
+        self.delora_A = nn.ParameterDict({})
+        self.delora_B = nn.ParameterDict({})
+        self.delora_alpha = nn.ParameterDict({})
+        # Store flag from kwargs (defaults to True if not provided)
+        self.use_residual_init = bool(kwargs.get("use_residual_init", True))
```

Comment:
Let's not do `kwargs.get(...)`, just put `use_residual_init: bool = True` into the `__init__` arguments.